Code/EnginePlugins/JoltPlugin/System/JoltWorldModule.cpp

```diff
@@ -762,5 +772,292 @@ void ezJoltWorldModule::UpdateConstraints()
+  }
+  else
+  {
+    pView->m_IncludeTags.Remove(tag);
+  }
+
+  if (cvar_JoltVisualizeGeometry)
+  {
+    const ezVec3 vCenterPos = pView->GetCamera()->GetCenterPosition();
+
+    DebugDrawGeometry(vCenterPos, cvar_JoltVisualizeDistance, ezPhysicsShapeType::Static, tag);
+    DebugDrawGeometry(vCenterPos, cvar_JoltVisualizeDistance, ezPhysicsShapeType::Dynamic, tag);
+    DebugDrawGeometry(vCenterPos, cvar_JoltVisualizeDistance, ezPhysicsShapeType::Query, tag);
+    DebugDrawGeometry(vCenterPos, cvar_JoltVisualizeDistance, ezPhysicsShapeType::Ragdoll, tag);
+    DebugDrawGeometry(vCenterPos, cvar_JoltVisualizeDistance, ezPhysicsShapeType::Trigger, tag);
+    DebugDrawGeometry(vCenterPos, cvar_JoltVisualizeDistance, ezPhysicsShapeType::Rope, tag);
+    DebugDrawGeometry(vCenterPos, cvar_JoltVisualizeDistance, ezPhysicsShapeType::Cloth, tag);
+  }
+
+  for (auto it = m_DebugDrawComponents.GetIterator(); it.IsValid();)
+  {
+    if (it.Value().m_uiLastSeenCounter < m_uiDebugGeoLastSeenCounter)
+    {
+      GetWorld()->DeleteObjectDelayed(it.Value().m_hObject);
+      it = m_DebugDrawComponents.Remove(it);
+    }
+    else
+    {
+      ++it;
+    }
+  }
+
+  for (auto it = m_DebugDrawShapeGeo.GetIterator(); it.IsValid();)
+  {
+    if (it.Value().m_uiLastSeenCounter < m_uiDebugGeoLastSeenCounter)
+    {
+      it = m_DebugDrawShapeGeo.Remove(it);
+    }
+    else
+    {
+      ++it;
+    }
+  }
+}
+
+void ezJoltWorldModule::DebugDrawGeometry(const ezVec3& vCenter, float fRadius, ezPhysicsShapeType::Enum shapeType, const ezTag& tag)
+{
+  const ezVec3 vAabbMin = vCenter - ezVec3(fRadius);
+  const ezVec3 vAabbMax = vCenter + ezVec3(fRadius);
+
+  JPH::AABox aabb;
+  aabb.mMin = ezJoltConversionUtils::ToVec3(vAabbMin);
+  aabb.mMax = ezJoltConversionUtils::ToVec3(vAabbMax);
+
+  JPH::AllHitCollisionCollector<JPH::TransformedShapeCollector> collector;
+
+  ezJoltBroadPhaseLayerFilter broadphaseFilter(shapeType);
+  JPH::ObjectLayerFilter objectFilterAll;
+  JPH::BodyFilter bodyFilterAll;
+
+  m_pSystem->GetNarrowPhaseQuery().CollectTransformedShapes(aabb, collector, broadphaseFilter, objectFilterAll, bodyFilterAll);
+
+  auto* pBodies = &m_pSystem->GetBodyInterface();
+
+  const int cMaxTriangles = 128;
+
+
+  ezStaticArray<ezVec3, cMaxTriangles * 3> positionsTmp;
+  positionsTmp.SetCountUninitialized(cMaxTriangles * 3);
+
+  ezStaticArray<const JPH::PhysicsMaterial*, cMaxTriangles> materialsTmp;
+  materialsTmp.SetCountUninitialized(cMaxTriangles);
+
+  ezHybridArray<ezVec3, cMaxTriangles * 3> positionsTmp2;
+  ezHybridArray<const JPH::PhysicsMaterial*, cMaxTriangles> materialsTmp2;
+
+  for (const JPH::TransformedShape& ts : collector.mHits)
+  {
+    DebugBodyShapeKey key;
+    key.m_uiBodyID = ts.mBodyID.GetIndexAndSequenceNumber();
+    key.m_pShapePtr = ts.mShape.GetPtr();
+
+    bool bExisted = false;
+    auto& geo = m_DebugDrawComponents.FindOrAdd(key, &bExisted).Value();
+    geo.m_uiLastSeenCounter = m_uiDebugGeoLastSeenCounter;
```

Comment:
Where is this last seen counter used? Shouldn't there be some code that deletes old entries after some time?

Reply:
At the end of DebugDrawGeometry() (see function above) the shapes are removed.